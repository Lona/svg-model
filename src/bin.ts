#!/usr/bin/env node

import fs from "fs";
import { convert } from "./index";

const [, , filename] = process.argv;

if (!filename) {
  console.log("No filename given");
  process.exit(1);
}

async function main() {
  const svgString = fs.readFileSync(filename, "utf8");

  const model = await convert(svgString);

  console.log(JSON.stringify(model, null, 2));
}

main();
