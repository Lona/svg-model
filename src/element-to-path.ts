import { SVGNode } from "./svg-types";

export default function elementToPath(element: SVGNode): string {
  const obj = require("element-to-path");
  const fn = obj && obj.__esModule ? obj["default"] : obj;
  return fn(element);
}
