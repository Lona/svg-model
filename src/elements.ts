import { Rect } from "./primitives";
import * as Commands from "./commands";
import parseCSSColor from "./parse-css-color";

export type LineCap = "butt" | "round" | "square";

export type FillRule = "nonzero" | "evenodd";

/**
 * We use different defaults in our model than the SVG spec.
 *
 * The SVG defaults are convenient for writing SVG files, but less convenient for drawing.
 * We get rid of the value 'none', instead using undefined (or no key).
 *
 * Model defaults:
 * - fill: undefined
 * - stroke: undefined
 * - strokeWidth: 0
 * - strokeLineCap: 'butt'
 *
 * SVG defaults:
 * - fill: 'black'
 * - stroke: 'none'
 * - strokeWidth: 1
 * - strokeLineCap: 'butt'
 */
export type Style = {
  fill?: string;
  fillRule: FillRule;
  stroke?: string;
  strokeWidth: number;
  strokeLineCap: LineCap;
};

export type Path = {
  id: string;
  commands: Commands.Command[];
  style: Style;
};

export type SVG = {
  children: Path[];
  viewBox?: Rect;
  metadata: {
    unsupportedFeatures: string[];
  };
};

function applyOpacity(color: string, opacity: number): string {
  const [r, g, b, a] = parseCSSColor(color) ?? [255, 255, 255, 1];

  if (opacity >= 1) return color;

  return `rgba(${r},${g},${b},${a * opacity})`;
}

const validateLineCap = (lineCap: string): LineCap => {
  if (lineCap !== "butt" && lineCap !== "round" && lineCap !== "square") {
    console.error(`Unsupported line cap value: ${lineCap}`);
    return "butt";
  }

  return lineCap;
};

/**
 * Convert SVG properties to our model format.
 *
 * We apply default values and convert them to a more convenient model for drawing.
 */
export const style = ({
  fill,
  fillOpacity,
  fillRule,
  stroke,
  strokeWidth,
  strokeLineCap,
  strokeOpacity,
}: {
  fill?: string;
  fillOpacity?: number;
  fillRule?: string;
  stroke?: string;
  strokeWidth?: number;
  strokeLineCap?: string;
  strokeOpacity?: number;
} = {}): Style => ({
  ...(fill !== "none" && {
    fill: applyOpacity(
      fill || "black",
      typeof fillOpacity === "number" ? fillOpacity : 1
    ),
  }),
  fillRule: fillRule === "evenodd" ? fillRule : "nonzero",
  ...(stroke &&
    stroke !== "none" && {
      stroke: applyOpacity(
        stroke,
        typeof strokeOpacity === "number" ? strokeOpacity : 1
      ),
    }),
  strokeWidth: strokeWidth != null ? strokeWidth : 1,
  strokeLineCap: strokeLineCap ? validateLineCap(strokeLineCap) : "butt",
});

export const path = (style: Style, commands: Commands.Command[]): Path => ({
  id: "",
  commands,
  style,
});

export const svg = (
  viewBox: Rect | undefined,
  unsupportedFeatures: string[]
): SVG => ({
  children: [],
  ...(viewBox && { viewBox }),
  metadata: {
    unsupportedFeatures,
  },
});
