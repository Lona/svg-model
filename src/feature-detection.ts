import { traverse } from "./traverse";
import { SVGRoot } from "./types/svg";

/**
 * All supported element names.
 */
export const SUPPORTED_ELEMENT_NAMES = [
  "circle",
  "defs",
  "desc",
  "g",
  "mask",
  "path",
  "polygon",
  "polyline",
  "rect",
  "svg",
  "title",
  "use",
];

/**
 * All supported element attributes.
 *
 * We could be more sophisticated and keep a map of which attributes can exist
 * on which element, but that doesn't seem necessary yet.
 */
export const SUPPORTED_ATTRIBUTE_NAMES = [
  "cx",
  "cy",
  "d",
  "fill-opacity",
  "fill-rule",
  "fill",
  "height",
  "href",
  "id",
  "points",
  "r",
  "rx",
  "ry",
  "stroke-linecap",
  "stroke-opacity",
  "stroke-width",
  "stroke",
  "transform",
  "viewBox",
  "width",
  "x",
  "xlink:href",
  "xmlns",
  "xmlns:xlink",
  "y",
];

function unique<T>(elements: T[]): T[] {
  return Array.from(new Set(elements).values());
}

export function getUnsupportedFeatures(
  root: SVGRoot
): { elements: string[]; attributes: string[] } {
  const elements: string[] = [];
  const attributes: string[] = [];

  traverse(root, (node) => {
    if (SUPPORTED_ELEMENT_NAMES.includes(node.name)) {
      if ("attributes" in node) {
        const attributeKeys = Object.keys(node.attributes);
        attributeKeys.forEach((key) => {
          if (!SUPPORTED_ATTRIBUTE_NAMES.includes(key)) {
            attributes.push(`${node.name}.${key}`);
          }
        });
      }
    } else {
      elements.push(node.name);
    }
  });

  return {
    elements: unique(elements),
    attributes: unique(attributes),
  };
}
