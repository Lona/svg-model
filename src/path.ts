import { point } from "./builders/primitives";
import { move, line, close, quadCurve, cubicCurve } from "./builders/commands";
import * as Commands from "./builders/commands";

const svgpath = require("svgpath");

export function convertCommand(segment: any[], x: number, y: number) {
  const command: string = segment[0];
  const parameters: number[] = segment.slice(1);

  switch (command) {
    case "M": {
      const [x, y] = parameters;
      return move(point(x, y));
    }
    case "L": {
      const [x, y] = parameters;
      return line(point(x, y));
    }
    case "H": {
      const [x] = parameters;
      return line(point(x, y));
    }
    case "V": {
      const [y] = parameters;
      return line(point(x, y));
    }
    case "Z": {
      return close();
    }
    case "Q": {
      let [qx1, qy1, qx2, qy2] = parameters;

      return quadCurve(point(qx2, qy2), point(qx1, qy1));
    }
    case "C": {
      const [x1, y1, x2, y2, x3, y3] = parameters;

      return cubicCurve(point(x3, y3), point(x1, y1), point(x2, y2));
    }
    default:
      console.error("Path segment not used:", segment);
      return null;
  }
}

export function convert(string: string, transform: string) {
  const parsed = svgpath(string);

  parsed.unarc();
  parsed.unshort();
  parsed.abs();

  if (transform) {
    parsed.transform(transform);
  }

  const drawCommands: Commands.Command[] = [];

  parsed.iterate((segment: any[], index: number, x: number, y: number) => {
    const command = convertCommand(segment, x, y);

    if (!command) return;

    drawCommands.push(command);
  });

  return drawCommands;
}
