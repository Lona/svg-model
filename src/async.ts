import SVGO from "svgo";
import { SVG, SVGWithoutQuadratics } from "./elements";
import { convert as convertSync } from "./index";
import { ConvertOptions } from "./node";

/**
 * Convert an SVG file string into a data model.
 *
 * @param svg
 * @param options {ConvertOptions}
 */
export async function convert(svg: string, options?: {}): Promise<SVG>;
export async function convert(
  svg: string,
  options?: { convertQuadraticsToCubics: true }
): Promise<SVGWithoutQuadratics>;
export async function convert(
  svg: string,
  options?: ConvertOptions
): Promise<SVG> {
  return convertSync(await optimize(svg), options);
}

async function optimize(svg: string) {
  const svgo = new SVGO({
    plugins: [
      { removeUselessDefs: true },
      { removeViewBox: false },
      { cleanupIDs: false },
      { convertColors: false },
    ],
  });
  const optimized = await svgo.optimize(svg);
  return optimized.data;
}
