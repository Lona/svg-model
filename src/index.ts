import { ConvertOptions, convertRoot } from "./node";
import { SVGRoot } from "./types/svg";
import { SVG, SVGWithoutQuadratics } from "./elements";
import SVGO from "svgo";

const parseSync: (string: string) => SVGRoot = require("svgson").parseSync;

/**
 * Synchronously convert an SVG file string into a data model.
 *
 * Many optimizations can't be performed, since we can't run `svgo` synchronously.
 *
 * @param svg {string}
 * @param options {ConvertOptions}
 */
export function convertSync(svg: string, options?: {}): SVG;
export function convertSync(
  svg: string,
  options?: { convertQuadraticsToCubics: true }
): SVGWithoutQuadratics;
export function convertSync(svg: string, options?: ConvertOptions): SVG {
  const root = parseSync(svg);
  return convertRoot(root, options);
}

/**
 * Convert an SVG file string into a data model.
 *
 * If running in node, we first optimize the SVG using SVGO.
 *
 * @param svg
 * @param options {ConvertOptions}
 */
export async function convert(svg: string, options?: {}): Promise<SVG>;
export async function convert(
  svg: string,
  options?: { convertQuadraticsToCubics: true }
): Promise<SVGWithoutQuadratics>;
export async function convert(
  svg: string,
  options?: ConvertOptions
): Promise<SVG> {
  return convertSync(
    process.env.PLATFORM === "web" ? svg : await optimize(svg),
    options
  );
}

async function optimize(svg: string) {
  const svgo = new SVGO({
    plugins: [
      { removeUselessDefs: true },
      { removeViewBox: false },
      { cleanupIDs: false },
      { convertColors: false },
    ],
  });
  const optimized = await svgo.optimize(svg);
  return optimized.data;
}

export * from "./elements";
export * from "./primitives";
export * from "./commands";
export * from "./style";
