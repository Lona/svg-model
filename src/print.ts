import { Command } from "./commands";
import { SVG } from "./elements";
import { Point } from "./primitives";

const round = (number: number, precision: number): number => {
  const base = Math.pow(10, precision);
  return Math.round((number + Number.EPSILON) * base) / base;
};

const stringifyPoint = (point: Point): string =>
  [round(point.x, 3), round(point.y, 3)].join(",");

const stringifyCommand = (command: Command) => {
  switch (command.type) {
    case "close":
      return "Z";
    case "move":
      return `M${stringifyPoint(command.to)}`;
    case "line":
      return `L${stringifyPoint(command.to)}`;
    case "quadCurve": {
      const parameters = [command.controlPoint, command.to].map(stringifyPoint);
      return `Q${parameters.join(" ")}`;
    }
    case "cubicCurve": {
      const parameters = [
        command.controlPoint1,
        command.controlPoint2,
        command.to,
      ].map(stringifyPoint);
      return `C${parameters.join(" ")}`;
    }
  }
};

type XMLElement = {
  name: string;
  attributes: Record<string, string>;
  children: XMLElement[];
};

function stringifyXMLElement({
  name,
  attributes,
  children,
}: XMLElement): string {
  const openingTagParts = [
    name,
    ...Object.entries(attributes).map(([name, value]) => `${name}="${value}"`),
  ];
  const openingTag = `<${openingTagParts.join(" ")}>`;
  const closingTag = `</${name}>`;

  const childrenString = children
    .map(stringifyXMLElement)
    .map((s) =>
      s
        .split("\n")
        .map((l) => "  " + l)
        .join("\n")
    )
    .join("\n");

  if (!childrenString) {
    return [openingTag, closingTag].join("");
  }

  return [openingTag, childrenString, closingTag].filter((s) => !!s).join("\n");
}

export function printSVG(svg: SVG) {
  const viewBox = svg.params.viewBox;
  const root: XMLElement = {
    name: "svg",
    attributes: {
      ...(viewBox && {
        width: `${viewBox.width}px`,
        height: `${viewBox.height}px`,
        viewBox: [viewBox.x, viewBox.y, viewBox.width, viewBox.height]
          .map((value) => value.toString())
          .join(" "),
      }),
      version: "1.1",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
    },
    children: svg.children.map((path): XMLElement => {
      const { commands, style } = path;

      return {
        name: "path",
        attributes: {
          fill: style.fill ?? "none",
          ...(style.stroke && { stroke: style.stroke }),
          ...(style.strokeWidth !== 1 && {
            "stroke-width": style.strokeWidth.toString(),
          }),
          ...(style.strokeLineCap && { "stroke-linecap": style.strokeLineCap }),
          d: commands.map(stringifyCommand).join(" "),
        },
        children: [],
      };
    }),
  };

  return `<?xml version="1.0" encoding="UTF-8"?>\n${stringifyXMLElement(root)}`;
}
