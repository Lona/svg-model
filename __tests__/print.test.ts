import fs from "fs";
import path from "path";
import { convert } from "../src";
import { printSVG } from "../src/print";

const readAsset = (name: string) => {
  const assetPath = path.join(__dirname, "../test/assets", name);
  return fs.readFileSync(assetPath, "utf8");
};

test("prints svg", () => {
  const asset = readAsset("check-circle.svg");
  const converted = convert(asset);
  const svg = printSVG(converted);
  expect(svg).toMatchSnapshot();
});

test("prints svg with quadratics", () => {
  const asset = readAsset("quadratics.svg");
  const converted = convert(asset);
  const svg = printSVG(converted);
  expect(svg).toMatchSnapshot();
});

test("prints svg without quadratics", () => {
  const asset = readAsset("quadratics.svg");
  const converted = convert(asset, {
    convertQuadraticsToCubics: true,
  });
  const svg = printSVG(converted);
  expect(svg).toMatchSnapshot();
});
