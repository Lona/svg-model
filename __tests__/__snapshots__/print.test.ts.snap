// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints svg 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<svg width=\\"24px\\" height=\\"24px\\" viewBox=\\"0 0 24 24\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\">
  <path fill=\\"#00C121\\" stroke-linecap=\\"butt\\" d=\\"M24,12 C24,18.627 18.627,24 12,24 C5.373,24 0,18.627 0,12 C0,5.373 5.373,0 12,0 C18.627,0 24,5.373 24,12 Z\\"></path>
  <path fill=\\"none\\" stroke=\\"#FFFFFF\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" d=\\"M6.5,12.6 L9.75,15.85 L17.25,8.35\\"></path>
</svg>"
`;

exports[`prints svg with quadratics 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<svg width=\\"1200px\\" height=\\"600px\\" viewBox=\\"0 0 1200 600\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\">
  <path fill=\\"none\\" stroke=\\"red\\" stroke-width=\\"5\\" stroke-linecap=\\"butt\\" d=\\"M200,300 Q400,50 600,300 Q800,550 1000,300\\"></path>
  <path fill=\\"black\\" stroke-linecap=\\"butt\\" d=\\"M210,300 C210,305.523 205.523,310 200,310 C194.477,310 190,305.523 190,300 C190,294.477 194.477,290 200,290 C205.523,290 210,294.477 210,300 Z\\"></path>
  <path fill=\\"black\\" stroke-linecap=\\"butt\\" d=\\"M610,300 C610,305.523 605.523,310 600,310 C594.477,310 590,305.523 590,300 C590,294.477 594.477,290 600,290 C605.523,290 610,294.477 610,300 Z\\"></path>
  <path fill=\\"black\\" stroke-linecap=\\"butt\\" d=\\"M1010,300 C1010,305.523 1005.523,310 1000,310 C994.477,310 990,305.523 990,300 C990,294.477 994.477,290 1000,290 C1005.523,290 1010,294.477 1010,300 Z\\"></path>
  <path fill=\\"#888888\\" stroke-linecap=\\"butt\\" d=\\"M410,50 C410,55.523 405.523,60 400,60 C394.477,60 390,55.523 390,50 C390,44.477 394.477,40 400,40 C405.523,40 410,44.477 410,50 Z\\"></path>
  <path fill=\\"#888888\\" stroke-linecap=\\"butt\\" d=\\"M810,550 C810,555.523 805.523,560 800,560 C794.477,560 790,555.523 790,550 C790,544.477 794.477,540 800,540 C805.523,540 810,544.477 810,550 Z\\"></path>
  <path fill=\\"none\\" stroke=\\"#888888\\" stroke-width=\\"2\\" stroke-linecap=\\"butt\\" d=\\"M200,300 L400,50 L600,300 L800,550 L1000,300\\"></path>
</svg>"
`;

exports[`prints svg without quadratics 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<svg width=\\"1200px\\" height=\\"600px\\" viewBox=\\"0 0 1200 600\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\">
  <path fill=\\"none\\" stroke=\\"red\\" stroke-width=\\"5\\" stroke-linecap=\\"butt\\" d=\\"M200,300 C333.333,133.333 466.667,133.333 600,300 C733.333,466.667 866.667,466.667 1000,300\\"></path>
  <path fill=\\"black\\" stroke-linecap=\\"butt\\" d=\\"M210,300 C210,305.523 205.523,310 200,310 C194.477,310 190,305.523 190,300 C190,294.477 194.477,290 200,290 C205.523,290 210,294.477 210,300 Z\\"></path>
  <path fill=\\"black\\" stroke-linecap=\\"butt\\" d=\\"M610,300 C610,305.523 605.523,310 600,310 C594.477,310 590,305.523 590,300 C590,294.477 594.477,290 600,290 C605.523,290 610,294.477 610,300 Z\\"></path>
  <path fill=\\"black\\" stroke-linecap=\\"butt\\" d=\\"M1010,300 C1010,305.523 1005.523,310 1000,310 C994.477,310 990,305.523 990,300 C990,294.477 994.477,290 1000,290 C1005.523,290 1010,294.477 1010,300 Z\\"></path>
  <path fill=\\"#888888\\" stroke-linecap=\\"butt\\" d=\\"M410,50 C410,55.523 405.523,60 400,60 C394.477,60 390,55.523 390,50 C390,44.477 394.477,40 400,40 C405.523,40 410,44.477 410,50 Z\\"></path>
  <path fill=\\"#888888\\" stroke-linecap=\\"butt\\" d=\\"M810,550 C810,555.523 805.523,560 800,560 C794.477,560 790,555.523 790,550 C790,544.477 794.477,540 800,540 C805.523,540 810,544.477 810,550 Z\\"></path>
  <path fill=\\"none\\" stroke=\\"#888888\\" stroke-width=\\"2\\" stroke-linecap=\\"butt\\" d=\\"M200,300 L400,50 L600,300 L800,550 L1000,300\\"></path>
</svg>"
`;
